File: mobile/android/app/src/main/java/com/turkuazlabs/telefonrehberi/mobile/services/SyncService.java
// # 📄 Dosya Yolu: C:/Projects/TelefonRehberi/mobile/android/app/src/main/java/com/turkuazlabs/telefonrehberi/mobile/services/SyncService.java
// # 📌 Amac: Android ile masaustu arasindaki iki yonlu rehber senkronizasyonu is kurallarini yonetir.
// # 📌 Service - Java
// Version: 2.37.1
// # Aciklama: Kalici sync UUID mappingi, guvenli LAN/HTTPS endpoint dogrulamasi, duplicate-safe pull ve idempotent push akisini uygular.
// # Bagimli Oldugu Katman: Service | Repository | Tool | Model | Language
package com.turkuazlabs.telefonrehberi.mobile.services;

import com.turkuazlabs.telefonrehberi.mobile.language.Messages;
import com.turkuazlabs.telefonrehberi.mobile.models.MobileContact;
import com.turkuazlabs.telefonrehberi.mobile.models.PullResult;
import com.turkuazlabs.telefonrehberi.mobile.repositories.DeviceContactRepository;
import com.turkuazlabs.telefonrehberi.mobile.repositories.SettingsRepository;
import com.turkuazlabs.telefonrehberi.mobile.tools.DesktopApiTool;
import com.turkuazlabs.telefonrehberi.mobile.tools.LanEndpointTool;

import java.util.List;

public final class SyncService {
    private final DeviceContactRepository contactsRepository;
    private final SettingsRepository settingsRepository;
    private final DesktopApiTool desktopApiTool;
    private final LanEndpointTool lanEndpointTool;

    public SyncService(
            DeviceContactRepository contactsRepository,
            SettingsRepository settingsRepository,
            DesktopApiTool desktopApiTool,
            LanEndpointTool lanEndpointTool
    ) {
        this.contactsRepository = contactsRepository;
        this.settingsRepository = settingsRepository;
        this.desktopApiTool = desktopApiTool;
        this.lanEndpointTool = lanEndpointTool;
    }

    public void saveConnection(String serverUrl, String token) {
        String validatedServerUrl = lanEndpointTool.validateBaseUrl(serverUrl);
        if (token == null || token.trim().isEmpty()) throw new IllegalArgumentException(Messages.TOKEN_REQUIRED);
        settingsRepository.saveConnection(validatedServerUrl, token);
    }

    public String serverUrl() { return settingsRepository.serverUrl(); }
    public String token() { return settingsRepository.token(); }

    public void testConnection() {
        requireSavedConnection();
        desktopApiTool.test(settingsRepository.serverUrl(), settingsRepository.token());
    }

    public PullResult pullDesktopContacts() {
        requireSavedConnection();
        List<MobileContact> desktopContacts = desktopApiTool.fetchContacts(settingsRepository.serverUrl(), settingsRepository.token());
        int added = 0;
        int skipped = 0;
        for (MobileContact contact : desktopContacts) {
            String mappedNativeId = settingsRepository.nativeIdForSyncUuid(contact.syncUuid());
            DeviceContactRepository.EnsureResult result = contactsRepository.ensureDesktopContact(contact, mappedNativeId);
            if (!contact.syncUuid().isBlank()) settingsRepository.saveSyncMapping(result.contactId(), contact.syncUuid());
            if (result.created()) added++; else skipped++;
        }
        return new PullResult(added, skipped);
    }

    public int pushDeviceContacts() {
        requireSavedConnection();
        List<MobileContact> deviceContacts = contactsRepository.findAll();
        String deviceId = settingsRepository.deviceId();
        for (MobileContact contact : deviceContacts) {
            String currentSyncUuid = settingsRepository.syncUuidForNativeId(contact.externalId());
            String canonicalSyncUuid = desktopApiTool.pushContact(
                    settingsRepository.serverUrl(), settingsRepository.token(), deviceId, currentSyncUuid, contact
            );
            if (!canonicalSyncUuid.isBlank()) settingsRepository.saveSyncMapping(contact.externalId(), canonicalSyncUuid);
        }
        return deviceContacts.size();
    }

    private void requireSavedConnection() {
        saveConnection(settingsRepository.serverUrl(), settingsRepository.token());
    }
}

File: mobile/ios/TurkuazTelefonRehberiIOS/services/SyncService.swift
// # 📄 Dosya Yolu: C:/Projects/TelefonRehberi/mobile/ios/TurkuazTelefonRehberiIOS/services/SyncService.swift
// # 📌 Amac: iOS ile masaustu arasindaki iki yonlu rehber senkronizasyonu is kurallarini yonetir.
// # 📌 Service - Swift
// Version: 2.38.1
// # 📌 Aciklama: Kalici sync UUID mappingi, duplicate-safe pull, idempotent push, Keychain hata propagasyonu ve acik hata mesajli guvenli LAN endpoint dogrulamasini uygular.
// # 📌 Bagimli Oldugu Katman: Service | Repository | Tool | Model | Language
import Foundation

@MainActor
final class SyncService {
    private let contactsRepository: DeviceContactRepository
    private let settingsRepository: SettingsRepository
    private let desktopApiTool: DesktopApiTool
    private let lanEndpointTool: LanEndpointTool

    init(
        contactsRepository: DeviceContactRepository,
        settingsRepository: SettingsRepository,
        desktopApiTool: DesktopApiTool,
        lanEndpointTool: LanEndpointTool = LanEndpointTool()
    ) {
        self.contactsRepository = contactsRepository
        self.settingsRepository = settingsRepository
        self.desktopApiTool = desktopApiTool
        self.lanEndpointTool = lanEndpointTool
    }

    func serverUrl() -> String { settingsRepository.serverUrl() }
    func token() -> String { settingsRepository.token() }

    func saveConnection(serverUrl: String, token: String) throws {
        let normalizedUrl: String
        do {
            normalizedUrl = try lanEndpointTool.normalizeBaseUrl(serverUrl)
        } catch LanEndpointError.required {
            throw SyncError.validation(Messages.serverRequired)
        } catch LanEndpointError.cleartextLocalOnly {
            throw SyncError.validation(Messages.serverHttpLocalOnly)
        } catch {
            throw SyncError.validation(Messages.serverUrlInvalid)
        }
        if token.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            throw SyncError.validation(Messages.tokenRequired)
        }
        try settingsRepository.saveConnection(serverUrl: normalizedUrl, token: token)
    }

    func testConnection() async throws {
        try requireSavedConnection()
        try await desktopApiTool.test(baseUrl: settingsRepository.serverUrl(), token: settingsRepository.token())
    }

    func pullDesktopContacts() async throws -> PullResult {
        try requireSavedConnection()
        try await contactsRepository.ensureAccess()
        let desktopContacts = try await desktopApiTool.fetchContacts(baseUrl: settingsRepository.serverUrl(), token: settingsRepository.token())
        var added = 0
        var skipped = 0
        for contact in desktopContacts {
            let mappedNativeId = settingsRepository.nativeId(forSyncUuid: contact.syncUuid)
            let result = try contactsRepository.ensureDesktopContact(contact, preferredContactId: mappedNativeId)
            if !contact.syncUuid.isEmpty { settingsRepository.saveSyncMapping(nativeContactId: result.contactId, syncUuid: contact.syncUuid) }
            if result.created { added += 1 } else { skipped += 1 }
        }
        return PullResult(added: added, skipped: skipped)
    }

    func pushDeviceContacts() async throws -> Int {
        try requireSavedConnection()
        try await contactsRepository.ensureAccess()
        let deviceContacts = try contactsRepository.findAll()
        let deviceId = settingsRepository.deviceId()
        for contact in deviceContacts {
            let currentSyncUuid = settingsRepository.syncUuid(forNativeId: contact.id)
            let canonicalSyncUuid = try await desktopApiTool.pushContact(
                baseUrl: settingsRepository.serverUrl(),
                token: settingsRepository.token(),
                deviceId: deviceId,
                syncUuid: currentSyncUuid,
                contact: contact
            )
            if !canonicalSyncUuid.isEmpty { settingsRepository.saveSyncMapping(nativeContactId: contact.id, syncUuid: canonicalSyncUuid) }
        }
        return deviceContacts.count
    }

    private func requireSavedConnection() throws {
        try saveConnection(serverUrl: settingsRepository.serverUrl(), token: settingsRepository.token())
    }
}

enum SyncError: LocalizedError {
    case validation(String)
    var errorDescription: String? {
        switch self { case let .validation(message): return message }
    }
}
